File: ios/Theme/Image.swift
import Foundation
import FaceTecSDK

public class Image: ImageStyle {
  private static let KEY: String = "image"
  private let style: Style
  private let target: NSDictionary?
  private var isPosition: Bool

  init() {
    self.style = Style()
    
    self.target = self.style.getTarget(Image.KEY)
    self.isPosition = false
    
    super.init(key: Image.KEY)
  }
  
  override init(target: NSDictionary?) {
    self.style = Style()
    
    self.target = target
    self.isPosition = false
    
    super.init(target: target)
  }
  
  override init(target: NSDictionary?, key: String) {
    self.style = Style()
    
    self.target = target
    self.isPosition = false
    
    super.init(target: target, key: key)
  }
  
  private func getBool(key: String) -> Bool {
    if !self.style.exists(self.target, key: key) {
      return true
    }

    return self.target?[key] as! Bool
  }

  public func getShowBranding() -> Bool {
    return self.getBool(key: "isShowBranding")
  }
  
  public func getHideForCameraPermissions() -> Bool {
    return self.getBool(key: "isHideForCameraPermissions")
  }
  
  public func getButtonPosition() -> CGRect {
    let cancelPosition = self.target?["cancelPosition"] as? NSDictionary
    let ios = cancelPosition?["ios"] as? NSDictionary
    
    if (cancelPosition == nil || ios == nil) {
      self.isPosition = false
      
      return .zero
    }
    
    let x = (ios?["x"] ?? 0) as! Int
    let y = (ios?["y"] ?? 0) as! Int
    let width = (ios?["width"] ?? 0.0) as! Double
    let height = (ios?["height"] ?? 0.0) as! Double
    
    let point = CGPoint(x: x, y: y)
    let size = CGSize(width: width, height: height)
    
    self.isPosition = true
    
    return CGRect(origin: point, size: size)
  }
  
  public func getButtonLocation() -> FaceTecCancelButtonLocation {
    let defaultLocation = FaceTecCancelButtonLocation.topRight

    let cancelLocation = self.isPosition ? "CUSTOM" : (self.target?["cancelLocation"] as? String) ?? ""

    switch cancelLocation {
    case "TOP_RIGHT":
      do {
        return FaceTecCancelButtonLocation.topRight
      }
    case "TOP_LEFT":
      do {
        return FaceTecCancelButtonLocation.topLeft
      }
    case "DISABLED":
      do {
        return FaceTecCancelButtonLocation.disabled
      }
    case "CUSTOM":
      do {
        return FaceTecCancelButtonLocation.custom
      }
    default:
      do {
        return defaultLocation
      }
    }
  }
}

File: android/src/main/java/com/azifacemobile/theme/Image.java
package com.azifacemobile.theme;

import android.graphics.Rect;

import com.azifacemobile.theme.abstracts.ImageStyle;
import com.azifacemobile.utils.Theme;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facetec.sdk.FaceTecCancelButtonCustomization;

import org.json.JSONException;
import org.json.JSONObject;

import javax.annotation.Nullable;

public class Image extends ImageStyle {
  private final static String KEY = "image";
  private final JSONObject target;
  private boolean isPosition;

  public Image(ReactApplicationContext context) {
    super(context, KEY);

    this.target = new Theme().getTarget(KEY);
    this.isPosition = false;
  }

  public Image(ReactApplicationContext context, JSONObject target) {
    super(context, target);

    this.target = target;
    this.isPosition = false;
  }

  public Image(ReactApplicationContext context, JSONObject target, String key) {
    super(context, target, key);

    this.target = target;
    this.isPosition = false;
  }

  private boolean getBoolean(String key, boolean defaultValue) {
    try {
      return this.target.getBoolean(key);
    } catch (NullPointerException | JSONException e) {
      return defaultValue;
    }
  }

  public boolean getShowBranding() {
    return this.getBoolean("isShowBranding", true);
  }

  public boolean getHideForCameraPermissions() {
    return this.getBoolean("isHideForCameraPermissions", false);
  }

  @Nullable
  public Rect getButtonPosition() {
    try {
      final JSONObject cancelPosition = this.target.getJSONObject("cancelPosition");

      final int left = cancelPosition.getInt("left");
      final int top = cancelPosition.getInt("top");
      final int right = cancelPosition.getInt("right");
      final int bottom = cancelPosition.getInt("bottom");

      this.isPosition = true;

      return new Rect(left, top, right, bottom);
    } catch (NullPointerException | JSONException e) {
      this.isPosition = false;

      return null;
    }
  }

  public FaceTecCancelButtonCustomization.ButtonLocation getButtonLocation() {
    final FaceTecCancelButtonCustomization.ButtonLocation defaultLocation = FaceTecCancelButtonCustomization.ButtonLocation.TOP_RIGHT;
    try {
      final String cancelLocation = this.isPosition ? "CUSTOM" : this.target.getString("cancelLocation");

      switch (cancelLocation) {
        case "TOP_RIGHT":
          return FaceTecCancelButtonCustomization.ButtonLocation.TOP_RIGHT;
        case "TOP_LEFT":
          return FaceTecCancelButtonCustomization.ButtonLocation.TOP_LEFT;
        case "DISABLED":
          return FaceTecCancelButtonCustomization.ButtonLocation.DISABLED;
        case "CUSTOM":
          return FaceTecCancelButtonCustomization.ButtonLocation.CUSTOM;
        default:
          return defaultLocation;
      }
    } catch (NullPointerException | JSONException e) {
      return defaultLocation;
    }
  }
}
